### ISSUE TITLE
Allow making python functions which would use `this` when called from js code

### ISSUE THREAD

**Issue by caryoscelus** [brython-dev/brython#400]
In js you can write something like this:
```js
obj = {
    f: function(){
        this.do_something
    }
}
obj.f()
```
Suppose there is js code that you want to mix with python and which relies on this behaviour (expects to be able to call `obj.f`). Attempt to port `obj` creation to python would not work because python doesn't have implicit `this`.
I can see two possible solutions. First: create wrapper which would transform python function so that when it's called from js it would receive `this` as first argument.
```python
def f(self):
    self.do_something
obj = {
    'f': js_pass_this(f)
}
# or use js_pass_this as decorator
```
Second: create `get_this()` function.
```python
def f():
    get_this().do_something
obj = {
    'f': f
}
```
I think first solution is both better and easier to implement though.

**Comment by caryoscelus** [comment]
Something like this seems to work:

```js
function pass_this(f) {
    new_f = function(){
        var args = [this]
        for (var i = 0; i < arguments.length; ++i) {
            args.push(arguments[i])
        }
        f.apply(null, args)
    }
    new_f.$infos = f.$infos
    return new_f
}
pass_this.$infos = {
        __name__:"foo",
        __defaults__ : [],
        __module__ : "__main__",
        __doc__: None,
        __annotations__: {},
        __code__:{
        __class__:$B.$CodeDict,
        co_argcount:1,
        co_filename:'js',
        co_firstlineno:1,
        co_flags:67,
        co_kwonlyargcount:0,
        co_name: "pass_this",
        co_nlocals: 1,
        co_varnames: ["f"]}
}
```
Obviously, this isn't proper way to introduce new brython function, but i don't know what is for such function anyway. I presume it should be placed in javascript module, not sure what is the best name though. Any thoughts?

**Comment by caryoscelus** [comment]
Actually, it works even without pretending to be brython function (the $infos part wasn't even working for me since i had it in context without None and $B). Also, `f.apply(null, args)` should be `return f.apply(null, args)`

**Comment by kikocorreoso** [comment]
@PierreQuentel what do you think about the @caryoscelus proposal?

@caryoscelus What would be the intended use of this? Use cases?

**Comment by caryoscelus** [comment]
Original use case is creating 'classes' with third-party js library. Specifically, i wanted to use brython with backbone and the function above works for that, but it should also work for other libraries using similar approach to making classes.

Example code:
```python
JSConstructor(backbone.Model.extend({'f':pass_this(f)}))
```

Or, more complex one, allowing to make backbone subclasses in python almost seamlessly:
```python
def backbone_class(js_cl):
    def wrap_attr(attr):
        if callable(attr) and not getattr(attr, '__backbone_class', False):
            return pass_this(attr)   # HERE, wrapping python function to pass into class constructor
        return attr
    
    class cl:
        def __new__(cls, *args):
            obj = JSConstructor(cls.get_js_constructor())(*args)
            return obj
        @classmethod
        def get_js_constructor(cls):
            ext = {name:wrap_attr(getattr(cls, name)) for name in dir(cls)}
            constructor = js_cl.extend(ext)
            constructor.__backbone_class = True
            return constructor
    return cl
```

**Comment by PierreQuentel** [comment]
Despite the explanations I honestly still don't understand what this is trying to solve. With this code

```
<script type="text/javascript">
obj = {
    f: function(){
        console.log(this)
    }
}
obj.f()
</script>

<script type="text/python">
from browser import window
window.obj.f()
</script>
```
I get the same result in the console for both scripts

**Comment by caryoscelus** [comment]
Because your example is not related: you're creating object inside js.

Somewhat close to your example would be this:

```html
<script type="text/javascript">
function test_call(obj) {
    obj.f()
}

js_obj = {
    f: function() {
        console.log(this);
    }
}

function pass_this(f) {
    new_f = function(){
        var args = [this]
        for (var i = 0; i < arguments.length; ++i) {
            args.push(arguments[i])
        }
        return f.apply(null, args);
    }
    new_f.$infos = f.$infos
    new_f.prototype = f.prototype
    for (var k in f) {
        new_f[k] = f[k]
    }
    
    return new_f
}
</script>

<script type="text/python">
from browser import window

def f(self):
    print(self)
obj = {
    'f': f
}
obj_fixed = {
    'f': window.pass_this(f)
}

window.test_call(window.js_obj)     # this works
window.test_call(obj_fixed)         # this works too
window.test_call(obj)               # this fails
</script>
```

**Comment by kikocorreoso** [comment]
Now I can better understand potential use cases.

I think this use can be quite specific and it could be solved externally to brython but, maybe, after testing and better justification a function (`javascriptify` or whatever) in the `javascript` module could be included as it seems it has potential powerful uses.

@PierreQuentel what do you think?

**Comment by PierreQuentel** [comment]
Since yesterday I am trying to imagine how I would write the documentation of this function, and I can't. How would you ?

**Comment by caryoscelus** [comment]
This function takes a (Brython) function `f` (which should have at least one argument, usually named `self`) and returns function `g`, which is supposed to be called from Javascript code and calls `f` with `this` as its first argument (other arguments are passed as usual). This allows to use Brython code for extending functionality of objects whose methods are called from JS. Note that calling resulting function from Brython code will result in first argument being set to an 'empty' `this`, usually `window` object.

Something like that. Some example similar to one i posted above should be also present. Perhaps it should use `pass_this` function as decorator, but for some reason that doesn't seem to work currently (#429).

**Comment by PierreQuentel** [comment]
This is a really complex definition ! It requires developers to dive into the code of Javascript packages to know that you must use this function to decorate Python functions (I have fixed #429)

Since the problem is to access `this` inside the Python function, I have tried another approach, which is much more simple to explain and might be more general that the one you propose : add a function `this()` in module __javascript__, that returns the current value of Javascript `this`.

I also modified the code to transform Python dictionaries into Javascript objects so that if a value is a function, it is called with `this` set to the Javascript object. For instance, Python dictionary `obj = {'f': f}` is transformed into Javascript object `obj = {f: g}` where g is called with `this` set to `obj`.

I have just committed the code ; with these changes this code works in both cases :

```
<script type="text/javascript">
function test_call(obj) {
    obj.f()
}

js_obj = {
    f: function() {
        console.log(this);
    }
}

</script>

<script type="text/python">
from browser import window
from javascript import this # new funtion in module javascript

def f():
    print('run f', this())

obj = {
    'f': f
}

window.test_call(window.js_obj)     # this works
window.test_call(obj)               # this works too
</script>
```

I'm not sure if it completely solves your problem but if it can be improved to solve it I think it would be a cleaner solution.
